morph_tool/morphdb.py

```diff
@@ -0,0 +1,257 @@
+        else:
+            morphologies = []
+
+        self.morphologies = list(morphologies)
+        self._known_morphologies = set(map(hash, self.morphologies))
+
+    def __iter__(self):
+        return iter(self.morphologies)
+
+    def remove_morphs(self, removed_morphs: Iterable[MorphInfo]) -> None:
+        '''Removes morphologies from the database.'''
+        removed_morphs = set(removed_morphs)
+        self.morphologies = [morph for morph in self.morphologies
+                             if morph.name not in removed_morphs]
+
+        self._known_morphologies = set(map(hash, self.morphologies))
+
+        for name in removed_morphs:
+            self.lineage.pop(name, None)
+
+    def add_morph(self, morph_info: MorphInfo) -> bool:
+        '''Add a morphology to the database.
+
+        If the name, mtype, layer combination already exists, then nothing is added
+
+        Return:
+            True if morphology added, False otherwise
+        '''
+        key = hash(morph_info)
+        if key in self._known_morphologies:
+            return False
+        self._known_morphologies.add(key)
+        self.morphologies.append(morph_info)
+
+        lineage = {'dendrite': morph_info.dendrite_donor, 'axon': morph_info.axon_donor}
+        lineage = {k: v for k, v in lineage.items() if v is not None}
+        if lineage:
+            self.lineage[morph_info.name] = lineage
+        return True
+
+    @property
+    def df(self):
+        '''Returns a pandas.DataFrame view of the data with the following columns:
+
+           'name': the morpho name (without extension)
+           'mtype': the mtype (without msubtype),
+           'msubtype': the msubtype (the part of the fullmtype after the ":")
+           'fullmtype': the full mtype (mtype:msubtype)
+           'layer': the layer (as a string)
+
+           # repair related columns
+           'use_axon': states that the morphology's axon can be used as a donor for axon grafting
+           'use_dendrites': states that this morphology can be used as a recipient for axon grafting
+           'axon_repair': flag to activate axon repair
+           'dendrite_repair': flag to activate dendrites repair
+           'basal_dendrite_repair': flag to activate basal dendrites repair (dendrite_repair
+                                    must be true as well)
+           'tuft_dendrite_repair': flag to activate tuft dendrites repair (dendrite_repair
+                                    must be true as well)
+           'oblique_dendrite_repair': flag to activate oblique dendrites repair (dendrite_repair
+                                    must be true as well)
+           'unravel': flag to activate unravelling
+           'axon_inputs': the list of morphologies whose axon can be grafted on this morphology
+
+           # deprecated
+           'use_for_stats': Legacy flag that was used to determine if an axon was suitable to be
+                            used as a axoninput
```

Comment:
Maybe this flag should not appear at all. Or maybe be hidden by default ?

Reply:
We'll need it to move away from `MorphologyTracker`, a better name would be helpful, to make its semantics clear.

How do you plan to make things hidden?

Reply:
I'm balancing between:

- `show_deprecated` flag in the function signature. It would default to False
- Not exposing this flag at all. But that could be annoying the day we want to inspect something about it.
- Always expose it and let to the use the responsability to use the columns they need.

Reply:
I'm leaning towards
> Always expose it and let to the use the responsability to use the columns they need.

As long as the meaning is well documented.  It has the same maintenance problems as `show_deprecated`, since new/old properties will be added/deleted over time, so even if you set `show_deprecated=True`, you'd need to test what is actually returned.

Reply:
Yes and we can expect people to use only the columns they know/need/have a reason to use. The 'deprecated' tag in the docstring should deter people to use it without a good reason to do so.